### PR TITLE
Improves the packages graphql api

### PR DESCRIPTION
**Pull request by dinakajoy**
This pull request will add a GraphiQL interface that can be used to fetch Opam Packages data using the following queries:

  ``` allPackages => contains, offset, limit ```
Here, contains, offset and limit are all optional parameters. Offset and limit are used to paginate the packages while contains filters out all packages with name that has the `contains` letter(s) in it.
 
 ``` package => name,  version ```
This returns the details of a specified package name and version. If the version is not specified, the latest package version is returned.

 ``` packgeByVersions => name,  from, upto ```
This returns all versions of the specified package name if from and upto are not specified or a range of the packages whose versions falls within the `from` (from the first package version if not specified) and `upto` (to the last package version if not specified)  versions specified.

### CODE REVIEW

File: test/ocamlorg_web/graphql_test.ml [hunk 1]
```diff
@@ -145,48 +139,60 @@ let state_test () =
...
         ] )
-    ; ( "test that starts_with function compares two different strings"
+    ; ( "test that is_in_range function checks if a current version is within \
+         the start and last version range specified"
       , [ Alcotest.test_case
-            "and returns false"
+            "and returns true"
             `Quick
-            (start_with_test "gv" "Bt" false)
+            (is_in_range_test "3.12.0" "2.0.0" "5.9.2" true)
```
Comment (pitag-ha) [inline test/ocamlorg_web/graphql_test.ml:153]: Since you're using `is_in_range` on type `Package.Version.t`, it might be better to also test it on that type instead of testing it on `string`. What do you think?

**Review by pitag-ha — COMMENTED** [review]
Nice, these are great improvements. Thanks!

File: lib/ocamlorg_web/graphql.ml [hunk 2]
```diff
@@ -6,59 +6,125 @@ type package_info =
...
+let is_valid_params limit offset total_packages =
+  let is_valid_limit = limit >= 0 && limit <= total_packages in
+  let is_valid_offset = offset >= 0 && offset <= total_packages in
+  let res = is_valid_limit in
+  match res with
+  | false ->
+    "wrong_limit"
+  | _ ->
+    let res = is_valid_offset in
+    (match res with false -> "wrong_offset" | _ -> "true")
```
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:38]: > Have you considered returning a custom variant type instead of a string? What do you think about that?

It's a bit unclear what you meant here but if you can drop an example, I will appreciate it

[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:38]: > Why do you have limit <= total_packages? I think the highest possible limit is total_packages - offset (+/- 1), isn't it? It might also be an option to assume that it's ok to exceed the highest possible limit.

Yea, I think it is better as `limit  > total_packages` as you specified in your if statement. Thank you

File: lib/ocamlorg_web/graphql.mli [hunk 3]
```diff
@@ -5,42 +5,57 @@ type package_info =
...
+  -> string option
+  -> Package.state
+  -> (Package.t, string) result
+
+val package_versions_result
+  :  string
+  -> string option
+  -> string option
+  -> Package.state
+  -> (packages_success, string) result
```
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:57]: I have added it, thank you

File: test/ocamlorg_web/graphql_test.ml (hunk 1 above)
Comment (dinakajoy) [reply test/ocamlorg_web/graphql_test.ml:153]: I am noticing a mix up here, the new test is not supposed to have `starts_with_test` as it has been removed. I will update in the next push

File: test/ocamlorg_web/graphql_test.ml [hunk 4]
```diff
@@ -88,49 +89,42 @@ let packages : Package.t list =
...
+let packages_list_test contains offset limit () =
+  let state = Package.state_of_package_list packages in
+  let all_packages =
+    Ocamlorg_web.Graphql.packages_list contains offset limit packages state
+  in
+  let num_of_packages_returned = List.length all_packages in
+  Alcotest.(check int)
+    "returns all the packages"
+    (List.length packages)
+    num_of_packages_returned
```
[the remark this answers is not in the document]
Comment (dinakajoy) [reply test/ocamlorg_web/graphql_test.ml:119]: Yea, I will add `contains` test

File: test/ocamlorg_web/graphql_test.ml (hunk 1 above)
Comment (pitag-ha) [reply test/ocamlorg_web/graphql_test.ml:235]: I think that's ok: `starts_with_test` is in red, so that means it was removed. What I was referring to in the comment was the line 
```
(is_in_range_test "3.12.0" "2.0.0" "5.9.2" true)
```
In that line, you're testing the function `is_in_range` on the strings `"3.12.0"`, `"2.0.0"` and `"5.9.2"`. You [use it in the code](https://github.com/ocaml/v3.ocaml.org-server/pull/78/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR117-R120) on type `Paclage.Version.t` though (which is the right thing to do). So I was suggesting to also test it on type `Paclage.Version.t` instead of testing it on type string.

File: test/ocamlorg_web/graphql_test.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply test/ocamlorg_web/graphql_test.ml]: Ok, ping me when you've added that so that I can have a look.

File: lib/ocamlorg_web/graphql.ml [hunk 5]
```diff
@@ -6,59 +6,125 @@ type package_info =
...
+          upto
+      in
+      let package_list =
+        List.filter
+          (fun package ->
+            is_in_range
+              (Package.version package)
+              (Package.Version.of_string from)
+              (Package.Version.of_string upto))
+          packages
```
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:121]: From your emoji reaction I assume that you don't like that suggestion very much? I think the main point of PR reviews is to discuss the code and to discuss different ideas. So it's cool if you have a different opinion and prefer to keep it the way it is now. It would be very helpful if you could point out why you prefer to keep it the way it is now.

File: lib/ocamlorg_web/graphql.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: > It's a bit unclear what you meant here but if you can drop an example, I will appreciate it

I mean something along the lines
```
type params_validity = Valid_params | Wrong_limit | Wrong_offset
let is_valid_params limit offset total_packages =
  if limit < 0 || limit > total_packages then
    Wrong_limit
  else if offset < 0 || offset > total_packages then
    Wrong_offset
  else
    Valid_params
```

[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: > Yea, I think it is better as limit > total_packages as you specified in your if statement. Thank you

What I meant was that, if I understand the code below correctly, the highest possible limit is `total_packages - offset` (I might be off by 1 here), not `total_packages`. So I was suggesting either to change that bound to
```
limit > total_packages - offset (+/- 1)
```
or to completely leave it out since it might also make sense to allow a higher limit in the query and in that case to return all packages from the offset on.

File: lib/ocamlorg_web/graphql.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:121]: Not an angry emoji, it's a confused emoji. I meant that I did not understand it very well. I apologize for the misunderstanding

File: lib/ocamlorg_web/graphql.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml]: Offset will not always be 0. But I subtracted 1 from total_packages directly since it starts counting from 0. Thank you for pointing that out

File: test/ocamlorg_web/graphql_test.ml (hunk 1 above)
Comment (dinakajoy) [reply test/ocamlorg_web/graphql_test.ml]: Done

File: lib/ocamlorg_web/graphql.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: What I mean is that the upper bound for `limit` should either be 
```
limit > total_packages - offset (+/- 1)
```
or left out completely. I don't think it should be `limit > total_packages` or `limit > total_packages - 1` or anything like that. Do you understand why I'm saying that?

[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Also, what do you think about the [suggestion about the custom return type](https://github.com/ocaml/v3.ocaml.org-server/pull/78#discussion_r694727416)?

[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml]: Hi @pitag-ha please do you mean something like this?
```
  let is_valid_params limit offset total_packages =
  if limit > total_packages - offset (+/- 1) then
    "wrong_limit"
  else if offset < 0 || offset > total_packages - 1 then
    "wrong_offset"
  else
    "true"
  ```

[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml]: > Also, what do you think about the suggestion about the custom return type?

let me try it out now

File: lib/ocamlorg_web/graphql.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:121]: Ah ok, sorry, I misunderstood the emoji. What I mean is the following: currently, `is_in_range` is defined as
```
let is_in_range current_version from_version upto_version =
  current_version >= from_version && current_version <= upto_version
```
Instead you could define it as something like
```
let is_in_range current_version from_version upto_version =
  match from_version, upto_version with
  | None, None -> true
  | Some from_version, None ->  current_version >= from_version
  | None, Some upto_version -> current_version <= upto_version
  | Some from_version, Some uptp_version -> current_version >= from_version && current_version <= upto_version
```
Then you could directly pass `from` and `upto` as options to that function without unwrapping them:
```
            is_in_range
              (Package.version package)
              (Option.map Package.Version.of_string from)
              (Option.map Package.Version.of_string upto))
```
Just as a proposal.

File: lib/ocamlorg_web/graphql.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: I meant
```
if limit < 1 || limit > total_packages - offset (+/- 1) then
    Wrong_limit
```
or simply
```
if limit < 1 then
    Wrong_limit
```

[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Thanks for the changes!

**Comment by tmattio** [comment]
This LGTM, was there anything left to do before merging? cc @pitag-ha @dinakajoy

**Comment by pitag-ha** [comment]
From my side it looks good as well. The one comment of mine that's still open was just a proposal, it's not blocking.

File: lib/ocamlorg_web/graphql.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Nice! I think it's quite a bit easier to read now, thanks!

[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Just one last comment about this: if you wanted to, in your tests you could now also test the `is_in_range` function passing in `None` for `from` and/or `upto`. But it's not necessary. It's up to you if you want to do that :)

File: test/ocamlorg_web/graphql_test.ml [hunk 6]
```diff
@@ -145,47 +230,113 @@ let state_test () =
...
+         the start and last version range specified"
       , [ Alcotest.test_case
             "and returns false"
             `Quick
-            (start_with_test "gv" "Bt" false)
+            (is_in_range_test
+               (Package.Version.of_string "3.12.0")
+               (Some (Package.Version.of_string "10.0.0"))
+               (Some (Package.Version.of_string "2.97.2"))
+               false)
```
Comment (pitag-ha) [inline test/ocamlorg_web/graphql_test.ml:259]: @dinakajoy , this is was I'm referring to in [my comment above](https://github.com/ocaml/v3.ocaml.org-server/pull/78#discussion_r704276058): if you'd like to, you could now add another test for `is_in_range_test` testing something along the lines
```
            (is_in_range_test
               (Package.Version.of_string "3.12.0")
               None
               (Some (Package.Version.of_string "3.13.0"))
               true)
```
But as I was saying, it's not necessary to add that.

Comment (dinakajoy) [reply test/ocamlorg_web/graphql_test.ml:259]: Sure, so I added it this way: 
```
    ; ( "test that is_in_range function checks that current version is not \
         within the start and last version range specified"
      , [ Alcotest.test_case
            "and returns false"
            `Quick
            (is_in_range_test
               (Package.Version.of_string "3.12.0")
               (Some (Package.Version.of_string "10.0.0"))
               (Some (Package.Version.of_string "2.97.2"))
               false)
        ] )
    ; ( "test that is_in_range function checks that current version is within \
         the start and last version range specified"
      , [ Alcotest.test_case
            "and returns false"
            `Quick
            (is_in_range_test
               (Package.Version.of_string "3.12.0")
               None
               (Some (Package.Version.of_string "3.13.0"))
               true)
        ] )```

File: lib/ocamlorg_web/graphql.mli [hunk 7]
```diff
@@ -5,41 +5,85 @@ type package_info =
...
-    search the list of packages looking for the package with a name that begins
-    with [startswith] *)
-
-val get_single_package : Package.t list -> string -> Package.t option
-(** [get_single_package packages name] will search the list of packages looking
-    for the package with a name that matches [name], [None] is returned if no
-    package can be found. *)
+  -> Package.state
+  -> Package.t list
+(** [packages_list] function is a sub function of [all_packages_result] that
```
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.mli:47]: I think this documentation comment could be improved slightly as it doesn't explain what `contains` does, namely it first uses `Packages.search` to get a list of packages then applies the offset and limit. 

It might also be a little more ergonomic to make the `contains` an optional, labelled parameter so users don't have to explicitly pass `None` when they don't want it i.e.

```ocaml
val packages_list
  :  ?contains:string
  -> int
  -> int
  -> Package.t list
  -> Package.state
  -> Package.t list
```

File: lib/ocamlorg_web/graphql.ml [hunk 8]
```diff
@@ -271,25 +329,49 @@ let schema t : Dream.request Graphql_lwt.Schema.schema =
...
-          ~resolve:(fun _ () startswith offset limit ->
-            let packages = Package.all_packages_latest t in
-            let total_packages = List.length packages in
-            let limit =
-              match limit with None -> total_packages | Some limit -> limit
-            in
-            get_packages_result total_packages offset limit startswith packages)
-      ; field
+          ~resolve:(fun _ () contains offset limit ->
+            let all_packages = Package.all_packages_latest t in
```
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.ml:333]: Could this maybe be done inside the `all_packages_latest` function ? I don't mind if it isn't too much it just seems odd to me that `all_packages_result` requires both the state `t` and a list of all of the packages?

**Review by patricoferris — COMMENTED** [review]
Sorry for the delay in reviewing, thanks for the PR! Couple of comments, the most important one is the version comparing I think.

File: test/ocamlorg_web/graphql_test.ml [hunk 9]
```diff
@@ -145,47 +230,113 @@ let state_test () =
 let () =
   Alcotest.run
     "ocamlorg"
-    [ ( "test that starts_with function compares two similar strings"
+    [ ( "test that get_info function returns all the dependencies"
```
Comment (patricoferris) [inline test/ocamlorg_web/graphql_test.ml:233]: Also as a minor point, I appreciate the verbosity in these strings but they're perhaps a little too verbose 😆 . 

If you look at the Alcotest examples https://github.com/mirage/alcotest they are (a) smaller and (b) don't contain white spaces. This is pretty useful because the strings are used in the filenames of the output per test case so when something fails and you need to go digging in the output it makes it a little easier to get to in my experience.

File: lib/ocamlorg_web/graphql.mli [hunk 10]
```diff
@@ -5,41 +5,85 @@ type package_info =
...
+  | Wrong_limit
+  | Wrong_offset
+
+val get_info : (Package.Name.t * string option) list -> package_info list
+(** [get_info] function loops through the list part of Package.Info such as
+    Package.Info.dependencies and returns a list of name and constraints record
+    Package.Info record *)
+
+(* val is_in_range : 'a -> 'a -> 'a -> bool *)
+val is_in_range : 'a -> 'a option -> 'a option -> bool
```
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:30]: @patricoferris please do you mean something like this: 
```
   let is_in_range current_version from_version upto_version =
  let current_version = Package.Version.to_string current_version in
  match from_version, upto_version with
  | None, None ->
    true
  | Some from_version, None ->
    current_version >= (Package.Version.to_string from_version)
  | None, Some upto_version ->
    current_version <= (Package.Version.to_string upto_version)
  | Some from_version, Some upto_version ->
   current_version >= (Package.Version.to_string from_version) && current_version <= (Package.Version.to_string upto_version)
```
Or  you are suggesting we use `OpamVersion.compare v1 v2` to compare the versions

[the remark this answers is not in the document]
Comment (patricoferris) [reply lib/ocamlorg_web/graphql.mli:30]: The latter, use `OpamVersion.compare` but here it would be `Package.Version.compare` (they're the same function) see: https://github.com/ocaml/v3.ocaml.org-server/blob/d7899a5476555c24c97255d4dc784c07af795a0a/lib/ocamlorg/package.mli#L17-L26

Then you can update the `mli` to be something like :
```ocaml
val is_in_range : Package.Version.t -> Package.Version.t option -> Package.Version.t option -> bool
```

Alternatively, you could push this function back into the `lib/ocamlorg/package.ml` so others can use it too if that makes more sense. Either is fine by me :))

[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:30]: I think `val compare : t -> t -> int` returns an int but we need a bool for the filter function

File: lib/ocamlorg_web/graphql.mli (hunk 7 above)
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:47]: I have this now: 
```
   let packages_list ?(contains = None) offset limit all_packages t =
  match contains with
  | None ->
    List.filteri (fun i _ -> offset <= i && i < offset + limit) all_packages
  | Some letters ->
    List.filteri
      (fun i _ -> offset <= i && i < offset + limit)
      (Package.search_package t letters)
 ```

File: lib/ocamlorg_web/graphql.mli (hunk 10 above)
[the remark this answers is not in the document]
Comment (patricoferris) [reply lib/ocamlorg_web/graphql.mli:30]: Yes you will need to use the `int` it returns to reconstruct the logic you already have. For example consider this case:

```ocaml
| Some from_version, None ->
    current_version >= from_version
```

Would become `Package.Version.compare current_version from_version >= 0`. From the Stdlib docs on compare functions:

> compare x y returns 0 if x is equal to y, a negative integer if x is less than y, and a positive integer if x is greater than y.

File: lib/ocamlorg_web/graphql.mli (hunk 7 above)
Comment (patricoferris) [reply lib/ocamlorg_web/graphql.mli:47]: Yep this is great, you shouldn't need to explicitly write `?(contains = None)` it is the same as `?contains`.

File: test/ocamlorg_web/graphql_test.ml (hunk 9 above)
Comment (tmattio) [reply test/ocamlorg_web/graphql_test.ml:233]: FWIW, I tend to write pretty verbose test descriptions as well, probably a habit I got from Elixir (one test suite I have in mind is https://github.com/aaronrenner/phx_gen_auth/blob/master/priv/templates/phx.gen.auth/auth_test.exs). But in this instance, I'd agree with Patrick that this could be a bit shorter (e.g. "get_info returns all the dependencies")

File: lib/ocamlorg_web/graphql.mli (hunk 7 above)
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:47]: Yea, but it throws an error here 
```
     ~resolve:(fun _ () contains offset limit ->
            Lwt.return (all_packages_result contains offset limit t))
```

Comment (patricoferris) [reply lib/ocamlorg_web/graphql.mli:47]: You probably need: 

```ocaml
     ~resolve:(fun _ () contains offset limit ->
            Lwt.return (all_packages_result ~contains offset limit t))
```

Also just as a minor point, "throws an error" is usually used for your code running and an exception being thrown for example `10 / 0` throws (or raises) the `Division_by_zero` exception. Here, your code doesn't type check (or at least compile, they're basically the same in OCaml for a majority of the cases). The only reason I add this is I started looking for something in your code that raises an exception but then realised the problem :))

**Review by patricoferris — COMMENTED** [review]
All looking good to me, just one comment on the `?contains:string option` code. RWO does a good job of explaining this (probably better than I did) https://dev.realworldocaml.org/variables-and-functions.html#scrollNav-2-7

File: lib/ocamlorg_web/graphql.mli [hunk 11]
```diff
@@ -23,36 +23,42 @@ type params_validity =
...
 (** [is_valid_params] function is a sub function of [all_packages_result] that
-    takes [offset] [limit] [all_packages] and retruns an error message if limit
-    or ossfest is less than 0 or gretaer than [all_packages]*)
+    takes [offset] [limit] [all_packages] and retruns an error message of type
+    [params_validity]if limit or ossfest is less than 0 or gretaer than
+    [all_packages] *)
 
 val packages_list
-  :  string option
+  :  ?contains:string option
```
[the remark this answers is not in the document]
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.mli:45]: The challenge I am having is with the first function that `contains` was initially passed onto. I dropped the code on slack. Please help me go through it

**Review by patricoferris — APPROVED** [review]
Awesome, LGTM thanks @dinakajoy

**Review by tmattio — COMMENTED** [review]
This looks great, thanks for all your work @dinakajoy!
